### ISSUE TITLE
Docker engine swarm api service discovery

### ISSUE THREAD

**Issue by F21** [prometheus/prometheus#1766]
In docker 1.12, the docker engine will ship with swarm mode built in. This means that it is now possible to stand up a swarm cluster using a bunch of nodes with just docker installed. In addition, swarm mode will come with dns and health checks built-in, negating the need to run consul or some other service discovery mechanism. More info here: https://docs.docker.com/engine/swarm/

It would be nice if prometheus can directly use the new services API to discover services running in a swarm cluster: https://docs.docker.com/engine/reference/api/docker_remote_api_v1.24/#3-8-services

Perhaps the config option could be called `docker_swarm_sd`.

**Comment by brian-brazil** [comment]
> In addition, swarm mode will come with dns and health checks built-in, 

This will need to be bypassed for Prometheus service discovery.

We may want to wait for a release or two for this to stabilise before adding it, and ensuring there's sufficient interest to justify the maintenance effort of another SD.

**Comment by bvis** [comment]
This feature would be amazing. It would allow us to simplify some dependencies now we need to manage to maintain a dynamic Prometheus environment

**Comment by michaelharrer** [comment]
You could use dns_sd_configs.
Im running a global cadvisor service and a global prometheus/node-exporter service and can scrape all nodes with following config, using the
`tasks.<servicename>` feature of swarm mode discovery.

```
  - job_name: 'cadvisor'
    dns_sd_configs:
    - names:
      - 'tasks.cadvisor'
      type: 'A'
      port: 8080

  - job_name: 'node-exporter'
    dns_sd_configs:
    - names:
      - 'tasks.node-exporter'
      type: 'A'
      port: 9100
```

Its a workaround but functional.

**Comment by Cas-pian** [comment]
Any progress on this? I'm really looking forward to use this feature.
Thanks very much!

**Comment by genki** [comment]
@michaelharrer Unfortunately, there is no way to determine the node_exporter is running on which node. Only node_exporter itself knows it, but there's no option to provide it into its metrics (https://github.com/prometheus/node_exporter/issues/319)

**Comment by joonas-fi** [comment]
I just hacked in a proof-of-concept that syncs tasks from Swarm manager to Prometheus: https://github.com/function61/prometheus-docker-swarm

Current limitation is that Prometheus has to be running on a Swarm manager node.

**Comment by bvis** [comment]
@genki, @joonas-fi: I've updated the description of the image I created for getting the metrics: https://github.com/bvis/docker-prometheus-swarm. It's not perfect but it is very useful and the best I've seen until now.

@joonas-fi I'll try your solution when I get some time, probably it's a better alternative. And you don't need to have it running in a swarm manager node if you expose the metrics to the cluster thanks a a proxy. A similar approach to:

```
docker \
  service create \
  --mode global \
  --name docker-exporter \
  --network monitoring \
  --publish 4999 \
  basi/socat:v0.1.0
```
Or:
```
docker \
    service create --name docker-proxy \
    --network my-network\
    --mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock,readonly \
    --constraint 'node.role==manager' \
    rancher/socat-docker
```
That gets the docker swarm events and expose them in the `docker-proxy:2375` endpoint in the network.

**Comment by genki** [comment]
@bvis Thank you for great work, I have tested your work in advance ;) 
I have sent a PR #2312, this could solve this issue as long as the services are not in global mode but replication mode.

**Comment by genki** [comment]
@bvis I have implemented your second suggestion https://github.com/genki/prometheus/commit/afde884ddcb6374869046d744b4bac4685157a59

This inject meta labels "__domain", "__service", "__task" and "__host" after the query execution time using docker API.

**Comment by bvis** [comment]
@genki Do you have a prometheus image ready for use? I've built your image but I think I still need to do more steps to include it in the project provided by @joonas-fi, or am I wrong?

**Comment by genki** [comment]
@bvis Thank you for reporting :)
The injection is only taken place at querying on time series, so you couldn't see it on console, but I thought it is sufficiant.
I think this implementation is too optimized for docker users.
It is nice if there is more generalized and sophisticated way to pass meta data about sample source.

**Comment by joonas-fi** [comment]
@bvis: oh man, thanks for tip regarding creating a service that exposes the manager Docker socket (via constraint) over TCP, I didn't think of that as a way to loosen the requirement of running it on a Swarm node. :)

I will make the Docker URL given to Docker client configurable, as you pointed out!

I'm not sure what you mean by "as it tries to obtain the data from the ingress network". To my understanding the [ingress network is only for published ports and the routing mesh](https://docs.docker.com/engine/swarm/ingress/)? So if you publish the socat port, it will be public and therefore will be visible both from the ingress network AND the container's IP itself. Publishing seems unnecessary as the port shouldn't be public anyway (security issue) and you can reach the socat service just by its name without the port being public (provided that socat service and monitoring are on the same network), if I understand correctly. :)

I haven't given much thought/researched into services running on different networks (business services and monitoring on different networks). Currently my assumption is that everything's running on the same network. I'll document that caveat. It might be easy to implement, I just don't know it yet.

Just to be super clear to everyone, mine and @bvis 's projects achieve different things:

- Mine provides autodiscovery of your services running on Swarm that have metrics that Prom should scrape, but not **container or node metrics**
- @bvis 's solution provides container ("Docker container metadata") and node metrics, but not **metrics from the actual services** if you have services that expose Prom metrics

**Comment by bvis** [comment]
@genki The problem I see with your solution is that it does not allow to filter any query based on these values, then I cannot use it in my dashboard to get values from one or some hosts.

@joonas-fi You are right when you say that it's unnecessary to publish the ports of the exporters in the routing mesh, I have used it just for debugging purposes. At the moment I removed the "--publish" option in cadvisor and node-exporter your system started to scrape the values correctly. But for use it under different environments and conditions I suggest to you to implement the networking selection feature.

Another suggestion is that it would be better to split your "docker-prometheus-bridge" binary in another image to allow process isolation, with both services running in the same container some problems could come. Or try to add it in the prometheus itself.

On other hand my dashboard shows the container metrics cadvisor provides. And it's easy to extend it. It would be good to allow me to create issues in your project to do a better follow-up.

And a 3rd option: create a prometheus fork adding your both features: @joonas-fi and @genki. It could be very useful until the Prometheus project adds support to Docker swarm service discovery, or maybe they could accept your changes, it's one of the best things of the open source model. ;)

**Comment by genki** [comment]
@bvis Injected labels are only be usable for such as legend labels because it is not real labels in the scope of query.
Prometheus is using the labels as identifier of targets, so if inject something in it causes duplication of targets.
My motivation was just using injected labels for legend label in Grafana such like "{{__host}}".

**Comment by jmendiara** [comment]
Based on the Swarm discovery from @ContainerSolutions, I've coded that PoC that is working ok in our stagging env:
https://github.com/jmendiara/prometheus-swarm-discovery

Takes some of the great ideas from the original solution, but tries to fit best in a deployment where prometheus is executed in a swarm worker (dedicated) without mounting shared volumes between workers/masters (that is fairly complex in some cloud providers) and provides more swarm metadata

The original motivation was using the hostname from the worker as the `instance` label instead of the task endpoint that is get with @michaelharrer DNS solution
(see https://github.com/jmendiara/prometheus-swarm-discovery/blob/master/prometheus-configs/prometheus.yaml#L8-L9)

That client/server duality needed could be simplified if prometheus implements a generic `<remote_sd_config>` very similar to ` <file_sd_config>`, but gets the `static_config` array from a configured endpoint 

Please, let me know what do you think about this approach

**Comment by llitfkitfk** [comment]
https://github.com/vegasbrianc/prometheus/issues/91

**Comment by cuigh** [comment]
After several months of waiting, I have implemented a simple Swarm discovery in my fork rep, maybe you guys also need it:
https://github.com/cuigh/prometheus

Or download image directly
https://hub.docker.com/r/cuigh/prometheus/

## Configuration

For prometheus:

```yaml
- job_name: swarm
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  swarm_sd_configs:
  - api_server: http://docker-proxy:2375
    # api_version: 1.32
    # refresh_interval: 10s
    # timeout: 15s
```

For Swarm service, You can add several labels to control scraping:

* `prometheus.enable` - Required
* `prometheus.port` - Required
* `prometheus.network` - Required, 'host' or any other overlay network which prometheus and service both attached to
* `prometheus.path` - Optional, default is `/metrics`

**Comment by krjensen** [comment]
We really need this as well, could we get an indication from prometheus, if they want to include the functionality provided by @cuigh ?

**Comment by simonpasquier** [comment]
I'm closing the issue as unfortunately, we are currently not accepting new integrations.  [ContainerSolutions/prometheus-swarm-discovery](https://github.com/ContainerSolutions/prometheus-swarm-discovery) is listed in the Prometheus [documentation](https://prometheus.io/docs/operating/integrations/#file-service-discovery) to integrate Docker Swarm with the file service discovery.

We can only provide the stability and performance we want to provide if we can properly maintain the codebase. This includes, amongst others, test integrations in an automated and scalable fashion. For this reason, we are suggesting people integrate with the help of our generic interfaces. We have an integrations page on which integrations using our generic interfaces are listed.

Even if existing integrations can not be tested in an automated fashion, we will not remove them for reasons of compatibility. This also means that any additions we take on, or any changes to existing integrations we make or accept, will mean maintaining and testing these until at least the next major version, realistically even beyond that.

Feel free to question this answer on our [developer mailing list](https://groups.google.com/forum/#!forum/prometheus-developers), but be aware it's unlikely that you will get a different answer.

**Comment by bborysenko** [comment]
Be aware that [ContainerSolutions/prometheus-swarm-discovery](https://github.com/ContainerSolutions/prometheus-swarm-discovery) is not yet ready for production usage https://github.com/ContainerSolutions/prometheus-swarm-discovery/issues/9 due to file descriptors leaks.

**Comment by joonas-fi** [comment]
I updated my old proof of concept to use a better strategy: https://github.com/function61/promswarmconnect

Previously it used the file service discovery type to dynamically write the file to disk based on the info in Swarm. Its drawback was that we had to make changes to the Prometheus container, overriding the entrypoint and launching the file synchronizer binary AND Prometheus. This is not robust, because we would have had to logic to deal with either of the binaries crashing.

My new approach emulates the API of the existing Triton service discovery, so we can run the released Prometheus container from Docker Hub 100 % unchanged. All you have to do is write configuration for the Triton SD in the Prometheus config file.

**Comment by byrnedo** [comment]
nice :) I've not been doing much on swarm lately but after having used k8s I might start doing so again. K8s is reallllllly heavy. I like how light swarm is.

**Comment by joonas-fi** [comment]
I'm curious to hear about your experiences, since I'm interested about
learning about k8s! In what regard do you consider it "heavy" - is it
mentally hard to understand, or is it heavy on the server resources like
RAM and CPU? Both?

I'm currently running a super small startup, so having the ability to use
cheaper VMs with low overhead of the infrastructure layer is important to
me.

**Comment by SuperQ** [comment]
Docker Swarm Mode is popular enough that we can make an exception to the SD moratorium. We also previously discussed adding support for it according to @brian-brazil.

**Comment by brian-brazil** [comment]
I see no reason to make any exceptions, we continue to have issues maintaining what we already have. We also previously decided not to support it, and it sounds like what exists now is not what existed then.

**Comment by pdambrauskas** [comment]
Why continue supporting other SDs then?
Maybe it would be possible to have some kind of plugin system for that? Generating configuration files is not that comfortable. I do understand, that maintenance of all different kinds of SDs is not your focus, but maybe it would make sense to deprecate all custom SDs, and have some kind of entry point for pluggable integration. Something like Apache Kafka has for pluggable Serializers?

@brian-brazil , can you close this issue, if your team is not considering accepting implementation for it? I'm going with your suggested file-based configuration for now :)

**Comment by SuperQ** [comment]
I highly disagree with this all or nothing approach and the indefinite moratorium on features/improvements to one of our core features.

I think we should accept an implementation of swarm SD as long as it's reasonably well done.

**Comment by pdambrauskas** [comment]
It is not like "all or nothing", I see it more like responsibility split, you can keep existing integrations in default bundle of prometheus, but make it possible to add other integrations as plugins (Like I mentioned in kafka's case you just add your implementation to classpath and set config `value-serializer: com.github.YourImplementation`)

**Comment by joonas-fi** [comment]
I think the file-based approach suffers from the fact that you need to have the SD agent write files in the same container as Prometheus is in. Problems:

- You either have to build the custom SD on top of Prometheus' Docker image (which is what my first POC of promswarmconnect did), which forces the SD project to release their image for every Prometheus' release
- Or you have to schedule the SD container on the same host as the Prometheus' container and have them share a filesystem (deep, unclean coupling, might not play well with high availability setups..)

A cleaner approach would be to make the file SD support also http URLs from which the discovery file is fetched. Whether this is a new SD plugin or a change to the file SD plugin is beside the point.. this remote file (maybe JSON) support would get us very far, so the actual SD implementation would be a separate app/container.

**Comment by SuperQ** [comment]
There was talk a while back about a "standard discovery API" within the CNCF. AFAIK this work is stalled.

Having file_sd_configs support URLs is something I've proposed several times.

**Comment by simonpasquier** [comment]
> It is not like "all or nothing", I see it more like responsibility split, you can keep existing integrations in default bundle of prometheus, but make it possible to add other integrations as plugins (Like I mentioned in kafka's case you just add your implementation to classpath and set config `value-serializer: com.github.YourImplementation`)

@pdambrauskas unfortunately there is no practical option in Go otherwise I guess that pluggable SD would have been done a long time ago...

**Comment by darkl0rd** [comment]
Have you guys seen cuigh's post above? His implementation (https://github.com/cuigh/prometheus) is complete, fully integrated and confirmed working. Considering that he already did all the heavy lifting, why not simply integrate his implementation? By the looks of things he even seems more than happy to maintain it.. .

**Comment by SuperQ** [comment]
I think it would be great. @cuigh Would you be willing to open a PR to add it?

**Comment by hairyhenderson** [comment]
@SuperQ that was already rejected at #3687 😉

**Comment by WTFKr0** [comment]
Hey, just test the @cuigh fork and it fill my needs, but i prefed to stay in the main prom repo

As i understand, the good solution is to create a new custom sd mechanism like the example in here : https://github.com/prometheus/prometheus/tree/master/documentation/examples/custom-sd

With the code of @cuigh fork in

Anybody have started working on this ?

**Comment by SuperQ** [comment]
I would propose re-submitting #3687 as a new PR. We can take an official vote on the Prometheus developers list to decide if it's good enough to merge, rather than having one person on prometheus-team object.

**Comment by cuigh** [comment]
> Hey, just test the @cuigh fork and it fill my needs, but i prefed to stay in the main prom repo
> 
> As i understand, the good solution is to create a new custom sd mechanism like the example in here : https://github.com/prometheus/prometheus/tree/master/documentation/examples/custom-sd
> 
> With the code of @cuigh fork in
> 
> Anybody have started working on this ?

I still don't think it's a good idea to implement swarm_sd based on file_sd, unless HTTP is supported in file_sd.

**Comment by WTFKr0** [comment]
Yeah agree
But i think prometheus team implement this plugin mechanism to be a standard for all sd
As I understand, they want in the future move existing core sd out of prometheus binary too. So all sd will use the plugin mode
> By co-locating Prometheus and our new executable we can configure Prometheus to read the file_sd-compatible output of our executable, and therefore scrape targets from that service discovery mechanism. In the future this will enable us to move SD integrations out of the main Prometheus binary, as well as to move stable SD integrations that make use of the adapter into the Prometheus discovery package

See https://prometheus.io/blog/2018/07/05/implementing-custom-sd/

**Comment by SuperQ** [comment]
No, we don't want to remove SD from the core. We do want to make it easier to add new methods outside the core.

**Comment by WTFKr0** [comment]
OK

So who can resubmit PR for a vote ?

@cuigh I would like to improve a bit some code in your fork, can you enable issues on your fork so we can echange on that ?

**Comment by brian-brazil** [comment]
We discussed this at our monthly meeting today, the moratorium remains. Currently we're awaiting integration testing for a good swathe of our existing SDs, which any new SD would be expected to follow in the steps of.

**Comment by joonas-fi** [comment]
@brian-brazil could you then please add http support to the file SD (so the SD JSON can be fetched over HTTP), so we'd at least get a clean point of integration for adding SD agents running outside of Prometheus' container?

See use case of https://github.com/function61/promswarmconnect - this would be much cleaner if it could produce JSON compatible with the file SD agent!

**Comment by brian-brazil** [comment]
We have a moratorium on new SDs, and we already have a clean generic interface for integrations.

**Comment by joonas-fi** [comment]
> We have a moratorium on new SDs, and we already have a clean generic interface for integrations.

That interface is just passing complexity management to the users. With that interface I need to have a the SD binary (let's say promswarmconnect) running either:

1) Inside the Prometheus' container. In this case the SD plugin project needs to release a new version each time Prometheus releases a new version (the SD plugin project overlays its binary inside Prometheus' Docker image). This also requires a process supervisor, because now we're having unrelated processes running inside a single Docker image. I tried that in my first proof-of-concept of promswarmconnect project. This approach is not robust and creates unnecessary burden for SD plugin developers

2) Another container running on the same host as Prometheus runs on, and they have to share a filesystem. This is also far from clean. This also requires support from your orchestration layer allowing you to specify ("this container should always run on the same host as Prometheus"), unless you just want to schedule it manually and therefore forget automatic rescheduling if a host goes down..


I ask again, is all this complexity justified just because you don't want to add remote JSON support to the file SD? I can totally understand not wanting to add 4 138 different SD plugins you have to maintain, for trendiest service platform of the week, but we're asking for an olive branch here because what you're suggesting is far from elegant and especially not of a microservice philosophy which Prometheus in other regards so elegantly fits in.

TL;DR: generic HTTP based SD integration is the only elegant way we'll be able to build SD integrations outside of Prometheus' tree.

**Comment by brian-brazil** [comment]
> This also requires support from your orchestration layer allowing you to specify ("this container should always run on the same host as Prometheus")

The sidecar model is pretty standard, and not something you can really avoid if you're using Prometheus. We assume a POSIX system, and that includes processes being able to share filesystems, send each other signals etc.

> This approach is not robust and creates unnecessary burden for SD plugin developers

I've done it in the past, the bash scripting is a little finicky, but it's quite doable. Especially if you can use a non-ancient version of bash.

> generic HTTP based SD integration is the only elegant way we'll be able to build SD integrations outside of Prometheus' tree.

I disagree here, and there's many out there that build fine on what we have. Writing code and deploying it are separate concerns, and I don't think we should be adding features just because one particular deployment system happens to lack a basic feature.

**Comment by cuigh** [comment]
> @cuigh I would like to improve a bit some code in your fork, can you enable issues on your fork so we can echange on that ?

Already merged, thanks.

**Comment by webchi** [comment]
Docker Swarm rocks :metal:

**Comment by kz1000fan** [comment]
Looks like this issue has gone stale...I'm looking for mechanisms to implement metrics discovery for Swarm-hosted containers and came across this thread.  Any further progress/thoughts on whether this will be supported in the master branch?  Thanks.

**Comment by WTFKr0** [comment]
@kz1000fan I think it's no way here
Give a try to the https://github.com/cuigh/prometheus fork

**Comment by darkl0rd** [comment]
I am aware that there have been several discussions around this subject - has a decision since been made on whether to natively support swarm service discovery?

**Comment by SuperQ** [comment]
@darkl0rd Yes, we're willing to accept new discovery. The new rules are
* Find a core maintainer to be a sponsor.
* Write the code.

I'm happy to be the sponsor for the docker swarm discovery, but someone needs to write the code. :)

**Comment by darkl0rd** [comment]
@SuperQ there is a complete, working fork / pull request in here from @cuigh.

**Comment by brian-brazil** [comment]
There are several variants out there, however I've yet to see one which is of a standard where it could be inside Prometheus - for example no hardcoding of business logic.

**Comment by roidelapluie** [comment]
I am working on this in #7420
